### PR TITLE
feat(hooks): support updatedInput for transparent command rewriting

### PR DESCRIPTION
**Pull request by zoorpha**
## Related Issue

N/A — This is a small feature addition (34 lines) that enables transparent command rewriting via PreToolUse hooks, aligning with the existing hooks Beta documentation.

## Description

This PR adds support for `hookSpecificOutput.updatedInput` in the PreToolUse hook lifecycle, enabling hooks to **transparently rewrite tool arguments** without blocking and requiring a retry.

### Motivation

The current Kimi CLI hook system supports two actions from a PreToolUse hook:
- `allow` — execute the tool as-is
- `block` — prevent execution and return an error to the LLM

This works well for guardrails (e.g. "don't rm -rf /"), but is insufficient for integrations like [RTK](https://github.com/rtk-ai/rtk) that want to **rewrite** commands for token optimization (e.g. `git status` → `rtk git status`) without forcing the LLM to retry.

With this change, a hook can output:

```json
{
  "hookSpecificOutput": {
    "updatedInput": {
      "command": "rtk git status"
    }
  }
}
```

And Kimi will execute the tool with the rewritten arguments, transparently to the LLM.

### Changes

**`src/kimi_cli/hooks/runner.py`**
- Added `updated_input: dict[str, Any] | None = None` to `HookResult`
- Extended `run_hook()` to parse `hookSpecificOutput.updatedInput` from JSON stdout
- Deny decisions still take precedence over `updatedInput`

**`src/kimi_cli/soul/toolset.py`**
- After PreToolUse hooks fire, if any result contains `updated_input`, merge it into the tool arguments dict before `tool.call(arguments)`
- Guarded by `isinstance(arguments, dict)` to ensure type safety

**`tests/hooks/test_runner.py`**
- `test_json_updated_input_parsing` — verifies `updatedInput` is parsed into `HookResult.updated_input`
- `test_json_deny_takes_precedence_over_updated_input` — verifies `permissionDecision: deny` still blocks even when `updatedInput` is present

### Backward Compatibility

- Fully backward compatible. Hooks that don't output `updatedInput` behave exactly as before.
- The feature is opt-in on the hook side — no config changes required.

### Use Case: RTK Integration

[RTK](https://github.com/rtk-ai/rtk) (Rust Token Killer) is a CLI proxy that reduces LLM token consumption by 60–90% on common commands. With this PR + the [RTK PR #1391](https://github.com/rtk-ai/rtk/pull/1391), Kimi users can:

```bash
rtk init -g --kimi
```

And have all `Shell` tool calls transparently rewritten to their RTK equivalents.

### Testing

```bash
uv run pytest tests/hooks/test_runner.py -v
```

All 8 tests pass, including the 2 new ones:
```
tests/hooks/test_runner.py::test_json_updated_input_parsing PASSED
tests/hooks/test_runner.py::test_json_deny_takes_precedence_over_updated_input PASSED
```

The full hooks test suite (36 tests) also passes:
```bash
uv run pytest tests/hooks/ -v
```

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/MoonshotAI/kimi-cli/blob/main/CONTRIBUTING.md) document.
- [ ] I have linked the related issue, if any. (N/A for this small feature)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have run `make gen-changelog` to update the changelog. (Command timed out in this environment)
- [ ] I have run `make gen-docs` to update the user documentation. (Command timed out in this environment)

---

<a href="https://app.devin.ai/review/moonshotai/kimi-cli/pull/1935" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by zoorpha** [comment]
Thanks for the automated review. We've addressed the issues identified:

- Added runtime validation for `updatedInput` to prevent crashes from malformed hook output
- Added test coverage for invalid `updatedInput` edge cases (null, string, list)
- Updated hooks documentation (EN + ZH) to explain the transparent rewrite capability
- Committed the previously missing CHANGELOG entries

All tests and checks pass. Resolving this review.

File: src/kimi_cli/soul/toolset.py [hunk 1]
```diff
@@ -184,6 +184,12 @@ async def _call():
                             ),
                         )
 
+                # Apply updatedInput from hooks (transparent rewrite, e.g. RTK)
+                if isinstance(arguments, dict):
+                    for result in results:
+                        if result.updated_input:
+                            arguments.update(result.updated_input)
```
[the remark this answers is not in the document]
Comment (zoorpha) [reply src/kimi_cli/soul/toolset.py:191]: Addressed in the latest commit — we added an `isinstance` guard in `run_hook` so that only dict values are accepted as `updated_input`. Invalid values (string, array, null) are ignored, preserving the intended fail-open behavior. Thanks for the catch!

**Comment by zoorpha** [comment]
we already created another PR for the rtk tool 
https://github.com/rtk-ai/rtk/pull/1391

**Comment by senolcolak** [comment]
great to have this feature

**Comment by senolcolak** [comment]
seems like we cant help kimi improvements

**Comment by zoorpha** [comment]
I'll close the ticket
